src/blueapi/service/handler.py

```diff
@@ -80,7 +87,48 @@ def setup_handler(
     config: Optional[ApplicationConfig] = None,
 ) -> None:
     global HANDLER
-    handler = Handler(config)
+
+    provider = None
+    plan_wrappers = []
+
+    if config:
+        visit_service_client: VisitServiceClientBase
+        if config.env.data_writing.visit_service_url is not None:
+            visit_service_client = VisitServiceClient(
+                config.env.data_writing.visit_service_url
+            )
+        else:
+            visit_service_client = LocalVisitServiceClient()
+
+        provider = VisitDirectoryProvider(
+            data_group_name=config.env.data_writing.group_name,
+            data_directory=config.env.data_writing.visit_directory,
+            client=visit_service_client,
+        )
+
+        # Make all dodal devices created by the context use provider if they can
+        try:
+            from dodal.parameters.gda_directory_provider import (
+                set_directory_provider_singleton,
+            )
+
+            set_directory_provider_singleton(provider)
+        except ImportError:
+            logging.error(
+                "Unable to set directory provider for ophyd-async devices, "
+                "a newer version of dodal is required"
+            )
+
+        plan_wrappers.append(lambda plan: attach_metadata(plan, provider))
```

Comment:
should we not also call `set_directory_provider_singleton(provider)` from `blueapi.data_management.gda_directory_provider` also? Here, we're only setting the directory provider in dodal.

Reply:
This works because dodal is a library and blueapi is an application.

Blueapi runs in a Python runtime environment and imports dodal . That environment is the only "thing" in play. Blueapi and dodal essentially merge into one thing at runtime and that thing is called Python. That means we're not setting the global variable "in dodal", we're setting it in Python and it will be global for both dodal code and blueapi code. Does that make sense?

I appreciate that https://github.com/DiamondLightSource/blueapi/pull/315#discussion_r1371455708 may have caused some confusion because I had two global variables. We only need one and it needs to be in the dodal code because the dodal code is what depends on it.

Reply:
sure, that makes sense. Effectively, when we call `dodal.parameters.gda_directory_provider.set_directory_provider_singleton(provider)`, we set the global variable `_PROVIDER`. So, am I right in saying we don't need to define `_PROVIDER` in blueapi, we can just expect that at runtime (when we call this function) we will have a global variable, `_PROVIDER` (which is actually defined in dodal)?